File: algorithms/go/fft.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"math/cmplx"
	"os"
	"time"
)

func fft(a []complex128) {
	n := len(a)
	// bit-reversal
	j := 0
	for i := 1; i < n; i++ {
		bit := n >> 1
		for ; j&bit != 0; bit >>= 1 {
			j ^= bit
		}
		j ^= bit
		if i < j {
			a[i], a[j] = a[j], a[i]
		}
	}
	// butterfly
	for length := 2; length <= n; length <<= 1 {
		ang  := 2.0 * math.Pi / float64(length)
		wlen := complex(math.Cos(ang), math.Sin(ang))
		for i := 0; i < n; i += length {
			w := complex(1, 0)
			for jj := 0; jj < length/2; jj++ {
				u := a[i+jj]
				v := a[i+jj+length/2] * w
				a[i+jj]           = u + v
				a[i+jj+length/2]  = u - v
				w *= wlen
			}
		}
	}
}

func main() {
	reader := bufio.NewReader(os.Stdin)
	writer := bufio.NewWriter(os.Stdout)
	defer writer.Flush()

	var n int
	fmt.Fscan(reader, &n)
	a := make([]complex128, n)
	for i := range a {
		var x float64
		fmt.Fscan(reader, &x)
		a[i] = complex(x, 0)
	}

	start := time.Now()
	fft(a)
	ms := float64(time.Since(start).Nanoseconds()) / 1e6

	k   := n
	if k > 5 { k = 5 }
	sum := 0.0
	for i := 0; i < k; i++ { sum += cmplx.Abs(a[i]) }

	fmt.Fprintf(writer, "%.6f\n", sum)
	fmt.Fprintf(os.Stderr, "%.3f\n", ms)
}

File: algorithms/python/fft.py
import sys
import time
import math

def fft(a):
    n = len(a)
    # bit-reversal
    j = 0
    for i in range(1, n):
        bit = n >> 1
        while j & bit:
            j ^= bit
            bit >>= 1
        j ^= bit
        if i < j:
            a[i], a[j] = a[j], a[i]
    # butterfly
    length = 2
    while length <= n:
        ang  = 2 * math.pi / length
        wlen = complex(math.cos(ang), math.sin(ang))
        for i in range(0, n, length):
            w = complex(1, 0)
            for jj in range(length // 2):
                u = a[i + jj]
                v = a[i + jj + length // 2] * w
                a[i + jj]              = u + v
                a[i + jj + length // 2] = u - v
                w *= wlen
        length <<= 1

def main():
    data = sys.stdin.read().split()
    idx  = 0
    n    = int(data[idx]); idx += 1
    a    = [complex(float(data[idx + i]), 0) for i in range(n)]

    start = time.perf_counter()
    fft(a)
    ms = (time.perf_counter() - start) * 1000

    k   = min(5, n)
    s   = sum(abs(a[i]) for i in range(k))
    sys.stdout.write(f"{s:.6f}\n")
    sys.stderr.write(f"{ms:.3f}\n")

if __name__ == "__main__":
    main()